src/object/cli_obj.c

```diff
@@ -5024,66 +5035,40 @@ dc_obj_update(tse_task_t *task, struct dtx_epoch *epoch, uint32_t map_ver,
 	uint64_t		 dkey_hash;
 	int			 rc;
 
-	obj_task_init_common(task, DAOS_OBJ_RPC_UPDATE, map_ver, args->th,
-			     &obj_auxi, obj);
+	rc = obj_task_init(task, DAOS_OBJ_RPC_UPDATE, map_ver, args->th,
+			   &obj_auxi, obj);
+	if (rc != 0) {
+		obj_decref(obj);
+		D_GOTO(out_task, rc);
+	}
+
+	if (obj_auxi->tx_convert) {
+		obj_auxi->tx_convert = 0;
+		return dc_tx_convert(obj, DAOS_OBJ_RPC_UPDATE, task);
+	}
 
 	rc = obj_update_sgls_dup(obj_auxi, args);
 	if (rc) {
 		D_ERROR(DF_OID" obj_update_sgls_dup failed %d.\n", DP_OID(obj->cob_md.omd_id), rc);
-		obj_comp_cb(task, NULL);
-		goto out_task;
+		D_GOTO(out_task, rc);
 	}
 
 	rc = obj_rw_req_reassemb(obj, args, NULL, obj_auxi);
 	if (rc) {
 		D_ERROR(DF_OID" obj_req_reassemb failed %d.\n",
 			DP_OID(obj->cob_md.omd_id), rc);
-		obj_comp_cb(task, NULL);
-		goto out_task;
+		D_GOTO(out_task, rc);
 	}
 
 	dkey_hash = obj_dkey2hash(obj->cob_md.omd_id, args->dkey);
 	rc = obj_req_get_tgts(obj, NULL, args->dkey, dkey_hash,
 			      obj_auxi->reasb_req.tgt_bitmap, map_ver, false,
 			      false, obj_auxi);
-	if (rc != 0) {
-		if (rc != -DER_SHARDS_OVERLAP) {
-			int rc1;
-
-			task->dt_result = rc;
-			/* If the task needs to retry, it has to go through registered
-			 * object completion callback, otherwise tse_task_complete()
-			 * will complete the retry task, instead of retrying it.
-			 */
-			rc1 = tse_task_register_comp_cb(task, obj_comp_cb, NULL, 0);
-			if (rc1 != 0) {
-				D_ERROR("update task %p, register_comp_cb "DF_RC"\n",
-					task, DP_RC(rc1));
-				obj_comp_cb(task, NULL);
-			}
-			goto out_task;
-		}
+	if (rc != 0)
+		D_GOTO(out_task, rc);
 
-		if (obj_auxi->is_ec_obj && obj_auxi->req_reasbed) {
-			args->iods = obj_auxi->reasb_req.orr_uiods;
-			args->sgls = obj_auxi->reasb_req.orr_usgls;
-		}
```

Comment:
the above 4 lines was for the case that obj_rw_req_reassemb() executed above so it replaced args->iods/sgls as reassembled iods/sgls, but the dc_tx_convert() will reassemble then again, so here assign back the original user iods/sgls.
now it is removed, but I did not see where with such similar handling? 


In the new code, for example an new obj update task, it call obj_rw_req_reassemb() and then call obj_req_get_tgts() which returns -DER_SHARDS_OVERLAP, then its complete cb (obj_comp_cb) called that will set tx_convert flag, two things looks not clear to me -
1) how can that task be retried? in obj_comp_cb() it just set the tx_convert flag, but seems will not be retried?
2) when that task get retried, above L5045 will call dc_tx_convert() but the "args->iods/sgls" was set as reassembled iod/sgls, how can it reassemble it again in dc_tx_convert()?

Reply:
you are right, this should be set back before convert.